rktl_sim/nodes/simulator

```diff
@@ -358,5 +449,244 @@ class Simulator(object):
+    return spawn_bounds
+
+
+def generate_car_properties(self):
+    """
+    Set the car parameters based on those provided in the rospy directory (the properties of the car itself: length, max_speed, throttle, steering, simulate effort, etc)
+    - generates random max speed and steering throw by calling other functions
+    - used to generate these properties for each car
+    Create a list for each car depending on its properties:
+        Car id
+        Car pose publisher
+        Car odom publisher
+        Car effort subscribers
+        Car command subscribers
+    Publisher:
+        match_status
+    Service:
+        Sim_reset
+        Sim_create_car
+        sim_delete_Car
+    @return the generic car properties that are created for thsi specific run in a dictionary
+    """
+    # general paramters that are the same for every single car
+    # if too constrained in car length between what we detect and what it is - might learn differently
+    # use the get_sim_param in each function case: makes the most sense
+    car_properties = {'length': rospy.get_sim_param('/cars/length'),
+                      'max_speed': self.get_sim_param("/cars/throttle/max_speed"),
+                      'steering_throw': self.get_sim_param("/cars/throttle/max_steering_throw"),
+                      'throttle_tau': rospy.get_sim_param("/cars/throttle/tau"),
+                      'steering_rate': rospy.get_sim_param("/cars/steering/rate"),
+                      'simulate_effort': (self.mode == SimulatorMode.REALISTIC)}
+    self.car_properties = car_properties
+    return car_properties
+
+
+def delete_all_cars(self):
+    car_configs = rospy.get_param('~cars', [])
+    self.reset_lock.acquire()
+    for car_config in car_configs:
+
+        if 'name' not in car_config:
+            rospy.signal_shutdown('no "name" set for car config in sim')
+        car_name = car_config['name']
+        if 'randomize_pose' in car_config:
+            init_pose = None
+
+        self.sim.delete_car(self.car_ids[car_name])
+
+        # Delete all of hte cars subscribers
+        if car_name in self.car_effort_subs:
+            self.car_effort_subs[car_name].unregister()
+            del self.car_effort_subs[car_name]
+        if car_name in self.car_cmd_subs:
+            self.car_cmd_subs[car_name].unregister()
+            del self.car_cmd_subs[car_name]
+        # delete all of the cars publishers
+        if car_name in self.car_odom_pubs:
+            self.car_odom_pubs[car_name].unregister()
+            del self.car_odom_pubs[car_name]
+        if car_name in self.car_pose_pubs:
+            self.car_pose_pubs[car_name].unregister()
+            del self.car_pose_pubs[car_name]
+    # reseting vars to be safe
+    self.car_ids = {}
+    self.car_pose_pubs = {}
+    self.car_odom_pubs = {}
+    self.car_effort_subs = {}
+    self.car_cmd_subs = {}
+    self.reset_lock.release()
+
+    return DeleteCarResponse(True)
```

Comment:
To be used, this code needs to be in the service callback. Applied for the generate function too.

Reply:
the delete_all_cars function loops over all of the cars and deletes them. while the delete_car_cb only deletes 1 car. I have attached a screenshot comparing the two segments of code:
<img width="1646" alt="image" src="https://user-images.githubusercontent.com/58446594/177401637-322f1b90-8bec-4410-af4d-b887cc218229.png">
In addition, the delete_all_cars function remains unused as it was part of my initial solution of deleting all cars, and then regenerating them from scratch. Now, I just loop over the already generated cars and create them. I just left it in bc it might be useful in the future. 

The same applies for create_car_cb (creates only 1 car in a callback), and generate_all_cars (loops over all cars and generates them)
<img width="1645" alt="image" src="https://user-images.githubusercontent.com/58446594/177402122-77333ded-69f3-46d4-95be-a6f20a8dbdbe.png">

I only use generate_all_cars in the initialization of the simulator class. I don't use any of the callbacks in my changes, but I am sure that they are used in other places. Generate_all_cars is called only once. What should I do? I can add delete_all_cars_cb and generate_all_cb, as new call back functions, and I will loop over all of cars and delete them. I can also modify the current create_car_cb and delete_car_cb to delete/create all cars.

also out of curiosity, what is the req parameter in the create_car_cb and the delete_car_cb. Is it the path to the specific car?

sorry for the long message!

Reply:
If we aren't going to use the function, then I would recommend we delete it to keep things slim. Since deletion of each car only requires pulling out the list of IDs, I think we only need the `delete_car_cb` / `create_car_cb`. It feels good to keep things concise.

The `req` param is the service request message defined in `rktl_msgs`. We should add mypy type checks at some point.

Reply:
Im down to rework it this way. The only problem that I cannot work around is that the create/delete_car_cb uses are req which has things such as init_pose, init_yaw, and I'm certain that it has more attributes. My current function of generate_all_cars does all of this via looking over the car_configs list and doing the same functionality (ie for new init_positions, it just reads from rospy again), while most of it remain the same. (The same applies to the delete_car_cb, but I do agree that if we don't use it, I should remove it.

Therefore, my question is this: how do I rework create_car_cb so that I can loop over all of the cars in the generate_all_cars function without having a req but rather having the cars id if it has certain functionality that relies on the data held in a req?

Would it help to call bc I feel that I might be getting confused lol.

Reply:
That's partially my fault with changing my thoughts on the design. I'll try to layout what I now think is the best approach below:

We have two types of parameters: configuration-based and instance-based.

Configuration-based parameters determine the kinematics and dynamics of the sim. A few examples: car size, sim timestep, friction, etc.

Instance-based parameters determine the initial state of a new run. A few examples: car pose, ball pose, etc.

For any sim param, you should have the option of either specifiying a value:
```
car_length: X
``` 
Or a range:
```
car_length:
  max: X
  min: Y
```
If given a range, the sim will automatically select a value on each instance it retrieves its value.

On startup, we load both configuration and instance parameters and create an initial setup.

Throughout usage of the sim, users and other processes will want to create new instances. Previously, we used a service to do so, since it's human readable and understandable. This is less than ideal for the autonomy, since it may want to modify the instance value of each car, then reset to a new instance.

So I'm thinking our approach is this:

One global reset service. This starts a new clean slate (deletes old objects first), then reloads configuration params using `get_sim_param`, and sets up cars and other objects using the instance params (again using `get_sim_param`). This is perfect for other processes, since they can easily modify rospy params, then call the global reset. But now we've made it difficult for users who want to control instance params.

To solve this, we use the existing create / destroy car services as wrappers for param modification. So instead of the create car service actualing creating a new car in the sim object, it simply sets the instance param for us using `rospy.set_param()`.

How does this sound?

Reply:
That sounds like a great approach. I just want to clarify how a run of the simulator works to see if I am understanding it the same as you:

The sim is initialized, creating: 
configuration-based parameters ie car properties, time step, car turning radius etc via get_sim_param.
instance-based parameters ie car list, ball, field size etc
- the init function of the simulator creates configuration-based parameters while calling functions in the sim.py to create instance-based parameters
- uses the create_car_cb to create cars

In between runs, we keep the configuration-based properties constant, while we loop over the instance-based parameters and reset their values where we don't create and delete objects but rather reset their individual parameters (ie new spawn locations and new field size)
- the reset function handles the reset of instance parameters that are all stored mostly in the sim.py file.

at the end of the sim, we physically delete all of the instance-based parameters ( cars, ball objects), and all of the configuration-based parameters.
- with the delete_car_cb

However, I still don't understand how I use these functions with the req parameter if I do not have it at the init function and at the deletion step from the sim side. Would it be safe to keep my existing functions of delete_all_cars and create_all_cars as they loop over the cars present in the simulator through their IDs and not their reqs? Another solution might be to perform a type check on the delete_car_cb where I will pass an id into the req from the simulator side. a req is a dict while an id is not a dict lol. If the id is passed in then I will create a car off of it, without breaking any existing functionality that is caused by using a req for car creation. The same applies to car deletion.

Would this work, or am I still confused about the design?

Reply:
Todays call really helped me understand how everything functions! I completely reworked the reinitialization of the sim such that whenever the create_car_cb / delete_car_cb is invoked, the instance-based parameters are reset (params like props, engine, ball init position, car properties, car init pos and velocity etc..) while the config-based parameters are kept constant (urdf, rate, frame id, sim mode, etc...). In addition, I did not actually have to create/delete all of the objects, but I loop over all of them and reset their parameters, saving time. I also added two more callback functions (delete/create_all_cars) just because I already had the functions written and used in other places in the sim. Let me know if there is anything else that I am missing!